include/cuco/detail/hash_functions/murmurhash3.cuh

```diff
@@ -355,4 +355,206 @@ struct MurmurHash3_x64_128 {
+
+      k2 *= c2;
+      k2 = rotl32(k2, 16);
+      k2 *= c3;
+      h2 ^= k2;
+
+      h2 = rotl32(h2, 17);
+      h2 += h3;
+      h2 = h2 * 5 + 0x0bcaa747;
+
+      k3 *= c3;
+      k3 = rotl32(k3, 17);
+      k3 *= c4;
+      h3 ^= k3;
+
+      h3 = rotl32(h3, 15);
+      h3 += h4;
+      h3 = h3 * 5 + 0x96cd1c35;
+
+      k4 *= c4;
+      k4 = rotl32(k4, 18);
+      k4 *= c1;
+      h4 ^= k4;
+
+      h4 = rotl32(h4, 13);
+      h4 += h1;
+      h4 = h4 * 5 + 0x32ac3b17;
+    }
+    //----------
+    // tail
+    std::uint32_t k1 = 0;
+    std::uint32_t k2 = 0;
+    std::uint32_t k3 = 0;
+    std::uint32_t k4 = 0;
+
+    auto const tail = reinterpret_cast<uint8_t const*>(bytes) + nblocks * block_size;
+    switch (size & (block_size - 1)) {
+      case 15: k4 ^= static_cast<std::uint32_t>(tail[14]) << 16; [[fallthrough]];
+      case 14: k4 ^= static_cast<std::uint32_t>(tail[13]) << 8; [[fallthrough]];
+      case 13:
+        k4 ^= static_cast<std::uint32_t>(tail[12]) << 0;
+        k4 *= c4;
+        k4 = rotl32(k4, 18);
+        k4 *= c1;
+        h4 ^= k4;
+        [[fallthrough]];
+
+      case 12: k3 ^= static_cast<std::uint32_t>(tail[11]) << 24; [[fallthrough]];
+      case 11: k3 ^= static_cast<std::uint32_t>(tail[10]) << 16; [[fallthrough]];
+      case 10: k3 ^= static_cast<std::uint32_t>(tail[9]) << 8; [[fallthrough]];
+      case 9:
+        k3 ^= static_cast<std::uint32_t>(tail[8]) << 0;
+        k3 *= c3;
+        k3 = rotl32(k3, 17);
+        k3 *= c4;
+        h3 ^= k3;
+        [[fallthrough]];
+
+      case 8: k2 ^= static_cast<std::uint32_t>(tail[7]) << 24; [[fallthrough]];
+      case 7: k2 ^= static_cast<std::uint32_t>(tail[6]) << 16; [[fallthrough]];
+      case 6: k2 ^= static_cast<std::uint32_t>(tail[5]) << 8; [[fallthrough]];
+      case 5:
+        k2 ^= static_cast<std::uint32_t>(tail[4]) << 0;
+        k2 *= c2;
+        k2 = rotl32(k2, 16);
+        k2 *= c3;
+        h2 ^= k2;
+        [[fallthrough]];
+
+      case 4: k1 ^= static_cast<std::uint32_t>(tail[3]) << 24; [[fallthrough]];
+      case 3: k1 ^= static_cast<std::uint32_t>(tail[2]) << 16; [[fallthrough]];
+      case 2: k1 ^= static_cast<std::uint32_t>(tail[1]) << 8; [[fallthrough]];
+      case 1:
+        k1 ^= static_cast<std::uint32_t>(tail[0]) << 0;
+        k1 *= c1;
+        k1 = rotl32(k1, 15);
+        k1 *= c2;
+        h1 ^= k1;
+    };
+    //----------
+    // finalization
+    h1 ^= size;
+    h2 ^= size;
+    h3 ^= size;
+    h4 ^= size;
+
+    h1 += h2;
+    h1 += h3;
+    h1 += h4;
+    h2 += h1;
+    h3 += h1;
+    h4 += h1;
+
+    h1 = fmix32_(h1);
+    h2 = fmix32_(h2);
+    h3 = fmix32_(h3);
+    h4 = fmix32_(h4);
+
+    h1 += h2;
+    h1 += h3;
+    h1 += h4;
+    h2 += h1;
+    h3 += h1;
+    h4 += h1;
+
+    return {h1, h2, h3, h4};
+  }
+
+ private:
+  constexpr __host__ __device__ std::uint32_t rotl32(std::uint32_t x, std::int8_t r) const noexcept
+  {
+    return (x << r) | (x >> (32 - r));
+  }
```

Comment:
Probably not in this PR but this should be a common utility available to all hashers to avoid duplications. @sleeepyjack A free function or a member function in a base class?

Reply:
XXHash also uses a `rotl` function (both 32 and 64 bit version). Thus I'd say we move it to a free function similar to `sanitize_hash`.